### DISCUSSION TITLE
Usage of bpftrace in Large-Scale VM or Kubernetes Environments

### DISCUSSION THREAD

**Discussion by ANSANJAY** [bpftrace/bpftrace#2975] [Q&A]
Hello bpftrace Community,

I’m exploring the use of eBPF tools for monitoring and diagnostics in large-scale environments. Our focus is on extensive VM and Kubernetes (K8s) clusters, and we are particularly interested in the practical applications of bpftrace in such scenarios.

Background:
We are considering bpftrace for detailed system monitoring and diagnostics across a significant number of VMs and potentially within Kubernetes environments. Our objectives include performance analysis, troubleshooting, and operational insights.

Discussion Points:

I am reaching out to gather experiences and insights from the community, especially from those who have implemented bpftrace in large organizations or complex infrastructures. I am interested in:

Real-world Use Cases: How has bpftrace been utilized in large-scale VM or K8s environments? What were the primary objectives and use cases?

Challenges and Resolutions: What challenges, particularly regarding scalability, security, or resource management, were encountered in these deployments? How were they overcome?

Achieved Benefits: What benefits and improvements were observed by using bpftrace in these environments?

Recommendations and Best Practices: Could you share any best practices or advice for deploying and managing bpftrace at scale?

Purpose:
This discussion will aid us in understanding the feasibility and potential impact of integrating bpftrace into our infrastructure.

I appreciate any experiences, case studies, or advice you can share and look forward to a fruitful discussion.

Thank you!

**Comment by danobi** [comment]
Hi, thanks for your interest.

There's some prior art in open source world:

* https://github.com/iovisor/kubectl-trace
* https://blog.px.dev/distributed-bpftrace/
* https://www.andreasgerstmayr.at/talks/2020/devconf.cz/using_bpftrace_with_pcp_and_grafana.pdf

For proprietary, I have put distributed bpftrace into production on the majority of hosts in meta datacenters (although that was a few years ago so I'm unsure of the current status). Worked fine at the time. Not sure if I can say more (esp. since I don't work there anymore).

**Reply by ajor** [reply]
Still in use at Meta (thank you!) :)
